c: clean up warnings from MSVC

```
     9>C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\include\optional(83,17): warning C4244: 'initializing': conversion from '_Ty' to 'float', possible loss of data [C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\cpp-build\driver\postgresql\adbc-driver-postgresql-test.vcxproj]
                 with
                 [
                     _Ty=double
                 ]
       C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\apache-arrow-adbc-0.4.0\c\validation\adbc_validation.cc(1983,1): message : see reference to function template instantiation 'std::_Optional_destruct_base<_Ty,true>::_Optional_destruct_base<double>(std::in_place_t,double &&) noexcept' being compiled [C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\cpp-build\driver\postgresql\adbc-driver-postgresql-test.vcxproj]
                 with
                 [
                     _Ty=float
                 ]
       C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\apache-arrow-adbc-0.4.0\c\validation\adbc_validation.cc(1983,1): message : see reference to function template instantiation 'std::_Optional_destruct_base<_Ty,true>::_Optional_destruct_base<double>(std::in_place_t,double &&) noexcept' being compiled [C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\cpp-build\driver\postgresql\adbc-driver-postgresql-test.vcxproj]
                 with
                 [
                     _Ty=float
                 ]
       C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\include\optional(247,1): message : see reference to function template instantiation 'std::_Optional_construct_base<_Ty>::_Optional_construct_base<double>(std::in_place_t,double &&)' being compiled [C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\cpp-build\driver\postgresql\adbc-driver-postgresql-test.vcxproj]
                 with
                 [
                     _Ty=float
                 ]
       C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\include\optional(247,18): message : see reference to function template instantiation 'std::_Optional_construct_base<_Ty>::_Optional_construct_base<double>(std::in_place_t,double &&)' being compiled [C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\cpp-build\driver\postgresql\adbc-driver-postgresql-test.vcxproj]
                 with
                 [
                     _Ty=float
                 ]
       C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\apache-arrow-adbc-0.4.0\c\validation\adbc_validation.cc(953,1): message : see reference to function template instantiation 'std::optional<float>::optional<double,0>(_Ty2 &&) noexcept' being compiled [C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\cpp-build\driver\postgresql\adbc-driver-postgresql-test.vcxproj]
                 with
                 [
                     _Ty2=double
                 ]
       C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\apache-arrow-adbc-0.4.0\c\validation\adbc_validation.cc(953,21): message : see reference to function template instantiation 'std::optional<float>::optional<double,0>(_Ty2 &&) noexcept' being compiled [C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\cpp-build\driver\postgresql\adbc-driver-postgresql-test.vcxproj]
                 with
                 [
                     _Ty2=double
                 ]
       C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\apache-arrow-adbc-0.4.0\c\validation\adbc_validation.cc(996,3): message : see reference to function template instantiation 'void adbc_validation::StatementTest::TestSqlIngestNumericType<float>(ArrowType)' being compiled [C:\Users\runneradmin\AppData\Local\Temp\tmpAFA9.tmp\cpp-build\driver\postgresql\adbc-driver-postgresql-test.vcxproj]
```

--- a/c/driver/common/utils.c
+++ b/c/driver/common/utils.c
@@ -144,7 +144,9 @@ int StringBuilderAppend(struct StringBuilder* builder, const char* fmt, ...) {
 
     va_start(argptr, fmt);
     int ret = vsnprintf(builder->buffer + builder->size, n + 1, fmt, argptr);
-    assert(ret >= 0);
+    if (ret < 0) {
+      return errno;
+    }
 
     va_end(argptr);
   }

--- a/c/validation/adbc_validation.cc
+++ b/c/validation/adbc_validation.cc
@@ -950,8 +950,8 @@ void StatementTest::TestSqlIngestNumericType(ArrowType type) {
   if constexpr (std::is_floating_point_v<CType>) {
     // XXX: sqlite and others seem to have trouble with extreme
     // values. Likely a bug on our side, but for now, avoid them.
-    values.push_back(-1.5);
-    values.push_back(1.5);
+    values.push_back(static_cast<CType>(-1.5));
+    values.push_back(static_cast<CType>(1.5));
   } else {
     values.push_back(std::numeric_limits<CType>::lowest());
     values.push_back(std::numeric_limits<CType>::max());

